Visually separate pinned and unpinned streams while searching

preetmishra opened the pull request:
The first commit is a refactor that extracts `streams_view_divider()` from `streams_view()`. 

The `streams_view_divider()` is used in the second commit, in `update_streams`, for inserting a divider between search results.

Fixes #577.

kaustubh-nair commented:
@preetmishra This looks like a useful visual improvement :+1: 
One bug I noticed is two lines are displayed if I search for something then clear the text with backspace.

preetmishra commented:
@kaustubh-nair Thanks for the bug report. :+1:

I couldn't notice it initially it as I only have a few streams pinned and the stream list would update past the pinned streams whenever I cleared the search with <kbd>backspace</kbd> (#700 will fix this). However, with a bunch of streams pinned, it was pretty straightforward to reproduce.

I have pushed an update that will resolve the bug.

neiljp commented on zulipterminal/helper.py:
```diff
@@ -529,6 +530,21 @@ def format_string(names: List[str], wrapping_text: str) -> List[str]:
     return [wrapping_text.format(name) for name in names]
 
 
+def streams_view_divider() -> Any:
```
This is now basically a factory function which initializes a divider in another way - so maybe let's just make this a simple subclass?

preetmishra commented on zulipterminal/ui_tools/views.py:
```diff
@@ -245,6 +245,19 @@ def read_message(self, index: int=-1) -> None:
+    A custom urwid.Divider to visually separate pinned streams from the
+    unpinned streams.
+    """
+    def __init__(self, div_char: str='-', top: int=0, bottom: int=0) -> None:
```
Unlike the previous implementation, I have set the `div_char='-'` from the `__init__` instead. This shall help to maintain consistency (between `StreamsView` and `LeftColumnView` stream view dividers) if we ever wish to update the `div_char`.

preetmishra commented:
@neiljp Thanks for the review and the pointers.

I have extracted the divider as a subclass now and moved it into in `views.py`. The subclass does improve readability and make it cleaner and the code does fit better in `views.py`.

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -288,6 +288,25 @@ def update_streams(self, search_box: Any, new_text: str) -> None:
+                    first_unpinned_index = index
+                    break
+            if (first_unpinned_index != 0
+                    and first_unpinned_index != len(streams_display)):
```
Can we simplify this with `not in`?

neiljp commented on zulipterminal/ui_tools/views.py:
```diff
@@ -288,6 +288,25 @@ def update_streams(self, search_box: Any, new_text: str) -> None:
+                    and first_unpinned_index != len(streams_display)):
+                # Check whether a divider already exists.
+                if not isinstance(streams_display[first_unpinned_index],
+                                  urwid.Divider):
```
`StreamsViewDivider`?

Also, I remember you discussed the need for this section, but could you explain how it can happen in the comment, ie. what situation you're avoiding?

preetmishra commented:
@neiljp Thanks for the review. I have incorporated all the suggestions in the latest update.

neiljp commented:
@preetmishra Thanks for working on this! I'm merging this now :tada:
